File: src/ai_market_monitor/static/asset-logos.js
/* Drawing a coin's picture, once, for every surface in the product.
 *
 * The server decides *which* pictures exist for a coin — `core/asset_logos.py` is the
 * one owner of that. This walks them in order and draws the first one that genuinely
 * renders, leaving the letter monogram in place when none of them do.
 *
 * Two failures used to end here as a coin with no logo:
 *
 *   1. A stored picture was used *instead of* the catalogue rather than *before* it.
 *      A stored address that 404s or is withdrawn left `<img>` broken — no picture, no
 *      monogram, no fall-through to the catalogue entry that would have worked.
 *   2. Nothing waited to see whether the picture loaded. `img.src = url` is optimistic;
 *      the failure arrives later, as an event nobody was listening for.
 *
 * So a source is only accepted once the browser has actually decoded it.
 */
(() => {
  const catalogPrefix = "https://cdn.jsdelivr.net/npm/@web3icons/core@4.0.53/";
  /** moduleUrl -> object URL, or null once the catalog is known not to have it. */
  const catalogCache = new Map();

  async function importSvg(moduleUrl) {
    try {
      const module = await import(moduleUrl);
      if (typeof module.default === "string" && module.default.trim().startsWith("<svg")) {
        return URL.createObjectURL(new Blob([module.default], { type: "image/svg+xml" }));
      }
    } catch (_error) {
      // The catalog has no entry under this name. Not an error: most coins are not in it.
    }
    return null;
  }

  async function resolveCatalog(moduleUrl) {
    if (!moduleUrl || !moduleUrl.startsWith(catalogPrefix)) return null;
    if (catalogCache.has(moduleUrl)) return catalogCache.get(moduleUrl);
    const candidates = [moduleUrl];
    // Exchanges prepend a supply multiplier to some meme-coin tickers (1000SHIB,
    // 10000SATS, 1MBABYDOGE) that the icon catalog does not use in its own filenames.
    const symbol = moduleUrl.match(/\/([A-Z0-9]+)\.svg\.js$/)?.[1];
    const withoutMultiplier = symbol?.replace(/^(?:1000|10000|1M)(?=[A-Z])/, "");
    if (withoutMultiplier && withoutMultiplier !== symbol) {
      candidates.push(moduleUrl.replace(`/${symbol}.svg.js`, `/${withoutMultiplier}.svg.js`));
    }
    let result = null;
    for (const candidate of candidates) {
      result = await importSvg(candidate);
      if (result) break;
    }
    catalogCache.set(moduleUrl, result);
    return result;
  }

  /** An <img> that has finished loading, or null if this address does not draw. */
  function drawable(source, alt) {
    return new Promise((resolve) => {
      const image = new Image();
      image.decoding = "async";
      image.alt = alt;
      image.addEventListener("load", () => resolve(image), { once: true });
      image.addEventListener("error", () => resolve(null), { once: true });
      image.src = source;
    });
  }

  /**
   * Put the best available picture into `container`.
   *
   * `container` already holds the letter monogram, drawn by the server. It is only
   * replaced once a real picture has loaded, so a coin never flickers from letters to
   * emptiness — the fallback is the starting state, not a recovery step.
   */
  async function load(container, moduleUrl, symbol, directUrl = null, providerUrl = null) {
    // Stored picture first, then the market-data provider's, then the catalog. That is
    // the order `core/asset_logos.py` publishes and this must not invent its own: the
    // first two are pictures of *this coin*, the catalog is addressed by ticker alone,
    // so two coins sharing a ticker share its file.
    //
    // `directUrl` also accepts an array, so a caller holding the whole ordered list can
    // hand it over rather than picking one and losing the rest.
    const safe = (value) =>
      typeof value === "string" && value.startsWith("https://") ? value : null;
    const direct = Array.isArray(directUrl) ? directUrl : [directUrl];
    const sources = [...new Set(
      [...direct.map(safe), safe(providerUrl), moduleUrl].filter(Boolean),
    )];
    const identity = sources.join("|");
    if (!container || !sources.length || container.dataset.assetLogoLoaded === identity) return;
    container.dataset.assetLogoLoaded = identity;
    const alt = `${symbol || "Asset"} logo`;

    for (const source of sources) {
      const address = source === moduleUrl ? await resolveCatalog(moduleUrl) : source;
      if (!address) continue;
      const image = await drawable(address, alt);
      // Another load for this container started while this one was in flight — that one
      // owns the container now, so this result is dropped rather than drawn over it.
      if (container.dataset.assetLogoLoaded !== identity) return;
      if (image) {
        container.replaceChildren(image);
        container.dataset.assetLogoDrawn = "true";
        return;
      }
    }
    // Every source failed. The monogram is still there, which is the point of it.
  }

  function hydrate(root = document) {
    const selector =
      "[data-asset-logo-module], [data-asset-logo-url], [data-asset-logo-provider-url]";
    root.querySelectorAll(selector).forEach((container) => {
      load(
        container,
        container.dataset.assetLogoModule,
        container.dataset.assetLogoSymbol,
        container.dataset.assetLogoUrl,
        container.dataset.assetLogoProviderUrl,
      );
    });
  }

  window.HilalAssetLogos = { load, hydrate };
  hydrate();
})();

File: src/ai_market_monitor/core/asset_logos.py
"""Where a coin's picture comes from. One resolver, one address, one owner.

Two separate duplications lived here, and each one hid a different missing logo.

**The address.** The catalogue address was written out seven times — four in Python,
three more in templates and browser code — each with the version pin `@4.0.53` typed in
by hand. Raising the version meant finding all seven, and missing one would leave a page
quietly asking for a version that may not exist.

**The sources.** Worse, and the reason coins such as Mubarak showed three letters
instead of a logo: *which pictures exist for this coin* was answered independently in
eight places, and each one knew a different subset.

    lifecycle_dashboard      stored picture + catalogue
    sharia_passports         stored picture + catalogue
    dashboard (home)         stored picture + catalogue
    sharia_screening         stored picture only
    live_market_quotes       catalogue only
    dashboard_test (radar)   catalogue only, and `logo_url` hard-coded to None

A coin the catalogue has never heard of — every small or new token — has exactly one
picture: the one the platform stored on its own asset record when the identity was
verified. The three readers that never looked at that record could therefore never show
it. The same coin had a logo on one page and a monogram on the next, which is why this
reads as "some coins have no logo" rather than as a bug in a named place.

So the question is answered once, here, and every reader asks it. `asset_logo()` returns
*every* source in the order they should be tried, and `static/asset-logos.js` walks that
order in the browser, falling through to the next one whenever a picture genuinely fails
to load.

`tests/unit/test_asset_logo_catalogue.py` fails if a Python file writes the address out
again, if a front-end copy names a different version, or if any source is dropped.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

#: The icon catalogue this product draws coin logos from, pinned to one version.
LOGO_CATALOG = "https://cdn.jsdelivr.net/npm/@web3icons/core@4.0.53/dist/svgs/tokens/branded"

#: A ticker we are willing to put inside a URL.
#:
#: Not a guess about which coins exist — it is the boundary between a symbol and a
#: request. Anything outside it gets no catalogue address at all rather than a URL built
#: from unchecked text.
_SAFE_TICKER = re.compile(r"^[A-Z0-9]{1,24}$")

#: How many letters the fallback monogram shows.
#:
#: Three fits the round container at every size the product uses it at. Four overflows,
#: and two reads as an abbreviation of nothing.
_MONOGRAM_LETTERS = 3


@dataclass(frozen=True, slots=True)
class AssetLogo:
    """Every picture that exists for one coin, in the order to try them.

    ``monogram`` is not a failure state. It is the designed appearance of a coin with no
    picture anywhere, and it is always filled in, so a card never has nothing to draw.
    """

    ticker: str
    monogram: str
    #: The picture stored on the asset's own record when its identity was verified.
    #: Present for far more coins than the catalogue covers, and the only source for a
    #: small or newly listed token.
    image_url: str | None
    #: The shared icon catalogue, addressed by ticker. Covers the well-known coins.
    module_url: str | None

    @property
    def sources(self) -> tuple[str, ...]:
        """Every address to try, best first. Empty when only the monogram is available."""

        return tuple(url for url in (self.image_url, self.module_url) if url)


def asset_logo_module_url(symbol: str) -> str | None:
    """The catalogue entry for one coin, or ``None`` when the ticker cannot be one.

    ``symbol`` is the coin on its own — "SOL", not "SOL/USDT". The catalogue files are
    named in capitals, so the case is settled here rather than by each caller.
    """

    ticker = _ticker(symbol)
    return f"{LOGO_CATALOG}/{ticker}.svg.js" if ticker else None


def asset_logo(symbol: str, provider_ids: Any = None) -> AssetLogo:
    """Every picture for one coin, from its ticker and its stored record.

    ``provider_ids`` is the asset record's own ``provider_ids`` mapping — pass it
    whenever the caller has the record, and the picture stored there is preferred over
    the catalogue. A caller that has no record passes nothing and still gets the
    catalogue address and the monogram; what it must never do is pass ``None`` for the
    stored picture *while holding a record that has one*, which is how the Opportunities
    page ended up drawing letters for coins the platform had a logo for.
    """

    ticker = _ticker(symbol)
    return AssetLogo(
        ticker=ticker,
        monogram=(ticker or str(symbol or "").upper())[:_MONOGRAM_LETTERS] or "?",
        image_url=stored_logo_url(provider_ids),
        module_url=f"{LOGO_CATALOG}/{ticker}.svg.js" if ticker else None,
    )


def stored_logo_url(provider_ids: Any) -> str | None:
    """The picture recorded on an asset, if there is a usable one.

    Only ``https`` is accepted. A stored value that is blank, relative, or plain
    ``http`` is treated as absent rather than written into the page: a mixed-content
    image is blocked by the browser and shows as a broken picture, which looks worse
    than the monogram it replaced.
    """

    if not isinstance(provider_ids, dict):
        return None
    url = str(provider_ids.get("logo_url") or "").strip()
    return url if url.lower().startswith("https://") else None


def _ticker(symbol: Any) -> str:
    normalized = str(symbol or "").strip().upper()
    return normalized if _SAFE_TICKER.fullmatch(normalized) else ""
